abcde.1
.TH abcde 1
.SH NAME
abcde \- Grab an entire CD and compress it to Ogg/Vorbis, MP3, FLAC, Ogg/Speex, AAC, 
WavPack, Monkey's Audio (ape), MPP/MP+(Musepack), True Audio (tta), MP2 format
and/or AIFF (Audio Interchange File Format).
.SH SYNOPSIS
.B abcde
.I [options] [tracks]
.SH DESCRIPTION
Ordinarily, the process of grabbing the data off a CD and encoding it, then
tagging or commenting it, is very involved.
.BR abcde
is designed to automate this. It will take an entire CD and convert it into
a compressed audio format - Ogg/Vorbis, MPEG Audio Layer III (MP3), Free Lossless
Audio Codec (FLAC), Ogg/Speex, MPP/MP+(Musepack), M4A (AAC) wv (WavPack), 
Monkey's Audio (ape), Opus, True Audio (tta), MPEG Audio Layer II (MP2)
or AIFF (Audio Interchange File Format) format(s).
With one command, it will:
.TP
.B *
Do a CDDB and/or Musicbrainz query over the Internet to look up your CD or
use a locally stored CDDB entry, or read CD-TEXT from your CD if it's available
.TP
.B *
Download the album art appropriate for your music tracks with many
user configurable options for download and post download alterations
including automated embedding of the album art for some containers
.TP
.B *
Grab an audio track (or all the audio CD tracks) from your CD
.TP
.B *
Normalize the volume of the individual file (or the album as a single unit)
.TP
.B *
Compress to Ogg/Vorbis, MP3, FLAC, Ogg/Speex, MPP/MP+(Musepack), M4A, wv (WavPack), 
Monkey's Audio (ape), Opus format(s), True Audio (tta), MP2 or AIFF
(Audio Interchange File Format) all in one CD read
.TP
.B *
Comment or ID3/ID3v2 tag
.TP
.B *
Give an intelligible filename
.TP
.B *
Calculate replaygain values for the individual file (or the album as a single unit)
.TP
.B *
Delete the intermediate WAV file (or save it for later use)
.TP
.B *
Repeat until finished
.P
Alternatively,
.B abcde
can also grab a CD and turn it into a single FLAC file with an embedded
cuesheet which can be user later on as a source for other formats, and will be
treated as if it was the original CD. In a way,
.B abcde
can take a compressed backup of your CD collection.
.SH OPTIONS
.TP
.B \-1
Encode the whole CD in a single file. The resulting file uses the CD title
for tagging. If the resulting format is a flac file with an embedded cuesheet,
the file can be used as a source for creating other formats. Use "\-1 \-o
flac \-a default,cue" for obtaining such a file.
.TP
.B \-a [actions]
Comma-delimited list of actions to perform. Can be one or more of: cddb, cue,
read, getalbumart, embedalbumart, normalize, encode, tag, move, replaygain,
playlist, clean. Normalize and encode imply read. Tag implies cddb, read,
encode. Move implies cddb, read, encode, tag. Replaygain implies cddb, read,
encode, tag and move. Playlist implies cddb. embedalbumart implies getalbumart.
The default is to do all actions except cue, normalize, replaygain, getalbumart,
embedalbumart and playlist.
.TP
.B \-b
Enable batch mode normalization. See the BATCHNORM configuration variable.
.TP
.B \-B
Enable automatic embedding of album art with certain containers. As of
abcde 2.8.2 supported containers are mp3 (using eyeD3), flac (using
metaflac), m4a (using AtomicParsley), WavPack aka wv (using wvtag) and
experimental support for ogg (using vorbiscomment). This command line
option also calls the getalbumart function. Further details of album art
embedding using the embedalbumart function can be found in the abcde FAQ
document packaged with abcde.
.TP
.B \-c [filename]
Specifies an additional configuration file to parse. Configuration options
in this file override those in \fI/etc/abcde.conf\fR or \fI$HOME/.abcde.conf\fR.
.TP
.B \-C [discid]
Allows you to resume a session for
.I discid
when you no longer have the CD available (\fBabcde\fR will automatically resume if
you still have the CD in the drive). You must have already finished at
least the "read" action during the previous session.
.TP
.B \-d [devicename | filename]
CD\-ROM block device that contains audio tracks to be read. Alternatively, a
single-track flac file with embedded cuesheet.
.TP
.B \-D
Capture debugging information (you'll want to redirect this \- try 'abcde \-D
2>logfile')
.TP
.B \-e
Erase information about encoded tracks from the internal status file, to enable
other encodings if the wav files have been kept.
.TP
.B \-f
Force the removal of the temporary ABCDETEMPDIR directory, even when we have
not finished. For example, one can read and encode several formats, including
\'.ogg\', and later on execute a \'move\' action with only one of the given
formats. On a normal situation it would erase the rest of those encoded
formats. In this case, \fBabcde\fR will refuse to execute such command, except if \-f
is used.
.TP
.B \-g
Enable lame's \-\-nogap option.  See the NOGAP variable. WARNING: lame's
\-\-nogap disables the Xing mp3 tag.  This tag is required for mp3 players to
correctly display track lengths when playing variable-bit-rate mp3 files.
.TP
.B \-G
Download album art using the getalbumart function. This is best done with
CDDBMETHOD including "musicbrainz", and requires the installation of glyrc.
ImageMagick is an optional but highly recommended package. Further details
of getalbumart can be found in the abcde FAQ document packaged with abcde.
.TP
.B \-h
Get help information.
.TP
.B \-j [number]
Start [number] encoder processes at once. Useful for SMP systems. Overrides
the MAXPROCS configuration variable. Set it to "0" when using \fBdistmp3\fR to avoid
local encoding processes.
.TP
.B \-k
Keep the wav files after encoding.
.TP
.B \-l
Use the low-diskspace algorithm. See the LOWDISK configuration variable.
.TP
.B \-L
Use a local CDDB repository. See the CDDBLOCALDIR variable.
.TP
.B \-m
Create DOS-style playlists, modifying the resulting one by adding CRLF line
endings. Some hardware players insist on having those to work.
.TP
.B \-n
Do not query CDDB database. Create and use a template. Edit the template to
provide song names, artist(s), ...
.TP
.B \-N
Non interactive mode. Do not ask anything from the user. Just go ahead.
.TP
.B \-o [filetype][:filetypeoptions]
Select output type. Can be "vorbis" (or "ogg"), "mp3", "flac", "spx", "mpc", "m4a", 
"wav", "wv", "ape", "opus", "mka" or "aiff".  Specify a comma-delimited list of output types
to obtain all specified types.  See the OUTPUTTYPE configuration variable. One can
pass options to the encoder for a specific filetype on the command line separating
them with a colon. The options must be escaped with double-quotes.
.TP
.B \-O
When fetching album art, always ask the user if they have a URL or
local file to override whatever album art may have been automatically
selected and downloaded. See the OVERRIDEALBUMARTDOWNLOAD
configuration variable.
.TP
.B \-p
Pads track numbers with 0\'s.
.TP
.B \-P
Use Unix PIPES to read and encode in one step (USEPIPES). This disables multiple
encodings, since the WAV audio file is never stored in the disc. For more detail
on this option see the FAQ document in the source tarball.
.TP
.B \-r [hosts...]
Remote encode on this comma-delimited list of machines using \fBdistmp3\fR. See
the REMOTEHOSTS configuration variable.
.TP
.B \-s [fields...]
List, separated by commas, the fields to be shown in the CDDB parsed entries.
Right now it only uses "year" and "genre".
.TP
.B \-S [speed]
Set the speed of the CD drive. Needs CDSPEED and CDSPEEDOPTS set properly
and both the program and device must support the capability.
.TP
.B \-t [number]
Start the numbering of the tracks at a given number. It only affects the
filenames and the playlist. Internal (tag) numbering remains the same.
.TP
.B \-T [number]
Same as \-t but changes also the internal (tag) numbering. Keep in mind that
the default TRACK tag for MP3 is $T/$TRACKS so it is changed to simply $T.
.TP
.B \-U
Set CDDBPROTO to version 5, so that we retrieve ISO-8859-15 encoded CDDB
information, and we tag and add comments with Latin1 encoding.
.TP
.B \-v
Show the version and exit
.TP
.B \-V
Be more verbose. On slow networks the CDDB requests might give the
sensation nothing is happening. Add this more than once to make things
even more verbose.
.TP
.B \-x
Eject the CD when all tracks have been read. See the EJECTCD configuration
variable.
.TP
.B \-X [cue2discid]
Use an alternative "cue2discid" implementation. The name of the binary must be
exactly that. \fBabcde\fR comes with an implementation in python under the examples
directory. The special keyword "builtin" forces the usage of the internal
(default) implementation in shell script.
.TP
.B \-w [comment]
Add a comment to the tracks ripped from the CD. If you wish to use
parentheses, these will need to be escaped. i.e. you have to write
"\\(" instead of "(".
.TP
.B \-W [number]
Concatenate CD\'s. It uses the number provided to define a comment "CD #" and
to modify the numbering of the tracks, starting with "#01". For Ogg/Vorbis and
FLAC files, it also defines a DISCNUMBER tag.
.TP
.B \-z
DEBUG mode: it will rip, using \fBcdparanoia\fR, the very first second of each track
and proceed with the actions requested very quickly, also providing some
"hidden" information about what happens on the background. CAUTION: IT WILL
ERASE ANY EXISTING RIPS WITHOUT WARNING!
.TP
.B [tracks]
A list of tracks you want \fBabcde\fR to process. If this isn't specified, \fBabcde\fR
will process the entire CD. Accepts ranges of track numbers -
"abcde 1-5 7 9" will process tracks 1, 2, 3, 4, 5, 7, and 9.
.SH OUTPUT
Each track is, by default, placed in a separate file named after the track in a
subdirectory named after the artist under the current directory. This can be
modified using the OUTPUTFORMAT and VAOUTPUTFORMAT variables in your
\fIabcde.conf\fR. Each file is given an extension identifying  its compression
format, 'vorbis' for '.ogg', '.mp3', '.flac', '.spx', '.mpc', '.wav', '.wv',
\(aq.ape', '.opus', '.mka' or 'aiff'.
.SH CONFIGURATION
\fBabcde\fR sources two configuration files on startup - \fI/etc/abcde.conf\fR and
\fI$HOME/.abcde.conf\fR, in that order.
.PP
The configuration options stated in those files can be overridden by providing
the appropriate flags at runtime.
.PP
The configuration variables have to be set as follows:
.TP
.B VARIABLE=value
Except when "value" needs to be quoted or otherwise interpreted. If other
variables within "value" are to be expanded upon reading the configuration
file, then double quotes should be used. If they are only supposed to be
expanded upon use (for example OUTPUTFORMAT) then single quotes must be used.
.TP
All shell escaping/quoting rules apply.
.TP
Here is a list of options \fBabcde\fR recognizes:
.TP
.B CDDBMETHOD
Specifies the methods we want to use to retrieve the track
information. Three values are recognized: "cddb", "musicbrainz" and
"cdtext". List all the methods desired in a comma delimited list and
\fBabcde\fR will attempt them all, returning a list of all search
results. The "cddb" value needs the CDDBURL and HELLOINFO variables
described below. The "musicbrainz" value uses the Perl helper script
\fBabcde-musicbrainz-tool\fR to establish a conversation with the
Musicbrainz server for information retrieval. "cdtext" needs "icedax"
or "cdda2wav" to be installed.
.TP
.B CDDBURL
Specifies a server to use for CDDB lookups.
.TP
.B CDDBPROTO
Specifies the protocol version used for the CDDB retrieval of results. Version
6 retrieves CDDB entries in UTF-8 format.
.TP
.B HELLOINFO
Specifies the Hello information to send to the CDDB server. The CDDB
protocol requires you to send a valid username and hostname each time you
connect. The format of this is username@hostname.
.TP
.B CDDBLOCALDIR
Specifies a directory where we store a local CDDB repository. The entries must
be standard CDDB entries, with the filename being the DISCID value. Other
CD playing and ripping programs (like Grip) store the entries under \fI~/.cddb\fR
and we can make use of those entries.
.TP
.B CDDBLOCALRECURSIVE
Specifies if the CDDBLOCALDIR has to be searched recursively trying to find a
match for the CDDB entry. If a match is found and selected, and CDDBCOPYLOCAL
is selected, it will be copied to the root of the CDDBLOCALDIR if
CDDBLOCALPOLICY is "modified" or "new". The default "y" is needed for the local
CDDB search to work.
.TP
.B CDDBLOCALPOLICY
Defines when a CDDB entry should be stored in the defined CDDBLOCALDIR. The
possible policies are: "net" for a CDDB entry which has been received from the
net (overwriting any possible local CDDB entry); "new" for a CDDB entry which
was received from the net, but will request confirmation to overwrite a local
CDDB entry found in the root of the CDDBLOCALDIR directory; "modified" for a
CDDB entry found in the local repository but which has been modified by the
user; and "always" which forces the CDDB entry to be stored back in the root of
the CDDBLOCALDIR no matter where it was found, and no matter it was not edited.
This last option will always overwrite the one found in the root of the local
repository (if any). STILL NOT WORKING!!
.TP
.B CDDBCOPYLOCAL
Store local copies of the CDDB entries under the $CDDBLOCALDIR directory.
.TP
.B CDDBUSELOCAL
Actually use the stored copies of the CDDB entries. Can be overridden using the
"\-L" flag (if is CDDBUSELOCAL in "n"). If an entry is found, we always give
the choice of retrieving a CDDB entry from the internet.
.TP
.B SHOWCDDBFIELDS
Coma-separated list of fields we want to parse during the CDDB parsing.
Defaults to "year,genre".
.TP
.B OGGENCODERSYNTAX
Specifies the style of encoder to use for the Ogg/Vorbis encoder. Valid options
are \'oggenc\' (default for Ogg/Vorbis) and \'vorbize\'.
This affects the default location of the binary,
the variable to pick encoder command-line options from, and where the options
are given.
.TP
.B MP3ENCODERSYNTAX
Specifies the style of encoder to use for the MP3 encoder. Valid options are
\'lame\' (default for MP3), \'gogo\', \'bladeenc\', \'l3enc\' and \'mp3enc\'.
Affects the same way as explained above for Ogg/Vorbis.
.TP
.B FLACENCODERSYNTAX
Specifies the style of encoder to use for the FLAC encoder. At this point only
\'flac\' is available for FLAC encoding.
.TP
.B MP2ENCODERSYNTAX
Specifies the style of encoder to use for the MPEG-1 Audio Layer II (MP2) encoder.
At this point both \'twolame\' and \'ffmpeg\' are available for MP2 encoding.
.TP
.B SPEEXENCODERSYNTAX
Specifies the style of encoder to use for Speex encoder. At this point only
\'speexenc\' is available for Ogg/Speex encoding.
.TP
.B MPCENCODERSYNTAX
Specifies the style of encoder to use for MPP/MP+ (Musepack) encoder. At this
point we only have \'mpcenc\' available, from musepack.net.
.TP
.B AACENCODERSYNTAX
Specifies the style of encoder to use for M4A (AAC) encoder. We support \'fdkaac\'
as \'default\' as well as FFmpeg or avconv, neroAacEnc, qaac and fhgaacenc. If qaac,
refalac or FFmpeg / avconv are used it is also possible to generate Apple Lossless
Audio Codec (alac) files. Note that qaac, refalac and fhgaacenc are Windows applications
which require Wine to be installed.
.TP
.B TTAENCODERSYNTAX
Specifies the style of encoder to use for True Audio (tts) encoding. We
support \'tta\' as default but the older \'ttaenc\' can be used as well.
.TP
.B WVENCODERSYNTAX
Specifies the style of encoder to use for WavPack. We support \'wavpack\'
as \'default\' but \'ffmpeg'\ is the other option (Note that this is for
FFmpeg only as avconv does not have a native WavPack encoder).
.TP
.B APENCODERSYNTAX
Specifies the style of encoder to use for Monkey's Audio (ape). We support \'mac\',
Monkey's Audio Console, as \'default\'.
.TP
.B OPUSENCODERSYNTAX
Specifies the style of encoder to use for the Opus encoder. At this point only
\'opusenc\' is available for Opus encoding.
.TP
.B MKAENCODERSYNTAX
Specifies the style of encoder to use for the Matroska or mka container. At this
point only \'ffmpeg\' is available to utilise the mka container. Safe audio codecs
for mka include Vorbis, MP2, MP3, LC-AAC, HE-AAC, WMAv1, WMAv2, AC3, eAC3 and Opus.
See the FAQ document for more information.
.TP
.B AIFFENCODERSYNTAX
Specifies the style of encoder to use for Audio Interchange File Format (AIFF). 
At this point only \'ffmpeg\' is available to utilise the AIFF container. 
.TP
.B NORMALIZERSYNTAX
Specifies the style of normalizer to use.  Valid options are \'default\'
and \'normalize'\ (and both run \'normalize-audio\'), since we only support it,
ATM.
.TP
.B CDROMREADERSYNTAX
Specifies the style of cdrom reader to use. Valid options are \'cdparanoia\',
\'libcdio'\, \'debug\' and \'flac\'. It is used for querying the CDROM and 
obtain a list of valid tracks and DATA tracks. The special \'flac\' case is u
sed to "rip" CD tracks from a single-track flac file.
.TP
.B CUEREADERSYNTAX
Specifies the syntax of the program we use to read the CD CUE sheet. Right now
we only support \'mkcue\', but in the future other readers might be used.
.TP
.B KEEPWAVS
It defaults to no, so if you want to keep those wavs ripped from your CD,
set it to "y". You can use the "\-k" switch in the command line. The default
behaviour with KEEPWAVS set is to keep the temporary directory and the wav
files even you have requested the "clean" action.
.TP
.B PADTRACKS
If set to "y", it adds 0's to the file numbers to complete a two-number
holder. Useful when encoding tracks 1-9.
.TP
.B INTERACTIVE
Set to "n" if you want to perform automatic rips, without user intervention.
.TP
.B NICE VALUES
Define the values for priorities (nice values) for the different CPU-hungry
processes: encoding (ENCNICE), CDROM read (READNICE) and distributed encoder
with \fBdistmp3\fR (DISTMP3NICE).
.TP
.B PATHNAMES
The following configuration file options specify the pathnames of their
respective utilities: LAME, GOGO, BLADEENC, L3ENC, XINGMP3ENC, MP3ENC,
VORBIZE, OGGENC, FLAC, SPEEXENC, MPCENC, WAVEPACK, APENC, OPUSENC, ID3, EYED3,
METAFLAC, CDPARANOIA, CD_PARANOIA, CDDA2WAV, PIRD, CDDAFS, CDDISCID, CDDBTOOL, 
EJECT, MD5SUM, DISTMP3, VORBISCOMMENT, NORMALIZE, CDSPEED, MP3GAIN, VORBISGAIN, 
MPCGAIN, MKCUE, MKTOC, CUE2DISCID (see option "\-X"), DIFF, HTTPGET, GLYRC,
IDENTIFY, DISPLAYCMD, CONVERT, QAAC, WINE, FHGAACENC, ATOMICPARSLEY, FFMPEG,
TWOLAME, MID3V2, TTA and TTAENC.
.TP
.B COMMAND-LINE OPTIONS
If you wish to specify command-line options to any of the programs \fBabcde\fR uses,
set the following configuration file options: LAMEOPTS, GOGOOPTS, AIFFENCOPTS,
BLADEENCOPTS, L3ENCOPTS, XINGMP3ENCOPTS, MP3ENCOPTS, VORBIZEOPTS, WAVEPACKENCOPTS, APENCOPTS,
OGGENCOPTS, FLACOPTS, SPEEXENCOPTS, MPCENCOPTS, FAACENCOPTS, NEROAACENCOPTS, FDKAACENCOPTS, 
OPUSENCOPTS, ID3OPTS, EYED3OPTS, MP3GAINOPTS, CDPARANOIAOPTS, CDDA2WAVOPTS, PIRDOPTS, 
CDDAFSOPTS, CDDBTOOLOPTS, EJECTOPTS, DISTMP3OPTS, NORMALIZEOPTS, CDSPEEDOPTS, MKCUEOPTS, 
VORBISCOMMMENTOPTS, METAFLACOPTS, DIFFOPTS, FLACGAINOPTS, VORBISGAINOPTS, HTTPGETOPTS,
GLYRCOPTS, IDENTIFYOPTS, CONVERTOPTS, DISPLAYCMDOPTS, QAACENCOPTS, FHGAACENCOPTS,
ATOMICPARSLEYOPTS, FFMPEGENCOPTS, DAGRABOPTS, TWOLAMENCOPTS and TTAENCOPTS.
.TP
.B CDSPEEDVALUE
Set the value of the CDROM speed. The default is to read the disc as fast as
the reading program and the system permits. The steps are defined as 150kB/s
(1x).
.TP
.B ACTIONS
The default actions to be performed when reading a disc.

changelog
abcde 2.9.4-DEV

 -- Steve McIntyre <93sam@debian.org>  XXXXX

abcde 2.9.3

 * Replace non-portable sed code for generating offset list with some
   simple awk instead. Should now work on FreeBSD again.
 * Deal with bizarre cdda2wav behaviour when doing cdtext lookup - it
   writes the track information to *stdin*. Closes Issue #89:
   https://abcde.einval.com/bugzilla/show_bug.cgi?id=89
 * Add a warning in abcde.conf about mayb needing to install recode before
   using it in mungefilename()

 -- Steve McIntyre <93sam@debian.org>  Tue, 05 Feb 2018 09:55:12 +0000

abcde 2.9.2

 * Fix up lookup code to deal with spaces etc. in directory names
 * Cope with "stub" releases in Musicbrainz
 * Always call glyrc with LC_ALL=C. Thanks to Andreas Vögele for the
   report and the suggested fix. Closes Issue 83:
   https://abcde.einval.com/bugzilla/show_bug.cgi?id=50
 * Add a link to the CDDB data format in the abcde man page. Closes
   Issue 81: https://abcde.einval.com/bugzilla/show_bug.cgi?id=81
 * More fixups of track number padding. Thanks to John Straw for the
   patch.

 -- Steve McIntyre <93sam@debian.org>  Thu, 02 Aug 2018 23:07:12 +0800

abcde 2.9.1

 * Merged Debian packaging files updates
 * Fixups in the new Musicbrainz code:
   + Cope with not finding an entry for a CD!
   + Fix (again!) dealing with CDs with no release events

 -- Steve McIntyre <93sam@debian.org>  Fri, 09 Mar 2018 23:01:04 +0000

abcde 2.9

 * Add some documentation for the aged CD ripper dagrab. Thanks to 
   Teika Kazura for the notification and suggested documentation. This 
   closes Issue 50: https://abcde.einval.com/bugzilla/show_bug.cgi?id=50
 * Work by Matthias Andree <matthias.andree@gmx.de> to address the issue
   where abcde fails with accented characters from CD-TEXT. The issue and
   partial fix applied here documented in Issue 53:
   https://abcde.einval.com/bugzilla/show_bug.cgi?id=53
   and also in the abcde mailing list:
   https://lists.einval.com/pipermail/abcde-users/2017-January/000232.html
 * Allow for embedding of  album art downloaded by the getalbumart function. 
   Currently this is available for flac (using metaflac), mp3 (using eyed3),
    m4a (using AtomicParsley) and WavPack aka wv (using wvtag).
    This can be invoked in 3 ways:
     
     1. Use the commandline '-B' option (this will also call getalbumart)
     2. Use the commandline '-a embedalbumart' option to add to list of actions
     3. Use 'embedalbumart' in the 'ACTIONS' list in ~/.abcde.conf
     
   Still needs more development but it is perfectly usable at the moment!
 * Use md5 rather than md5sum under macOS. Thanks to JCount for the bug
   report and also the fix. This solves Issue 59:
   https://abcde.einval.com/bugzilla/show_bug.cgi?id=59
 * Support added for encoding with Audio Interchange File Format (AIFF). 
   Thanks to Massimo Villa for the feature request. FFmpeg is required
   for the encoding, the container and suffix are 'aiff'.
 * Allow selection of either cddb or musicbrainz from the command line:
 
  -Q   Select CDDBMETHOD from the command line. Choice is cddb or musicbrainz.
  
   Command line letters are fast running out but the 'Q' option quite neatly
   stands for 'Query'!
 * Allow for embedding with do_embedalbumart() for single track encodes when
   OUTPUTFORMAT and ONETRACKOUTPUTFORMAT are different. Thanks to Ashley Gittins
   for the bug report. This closes Issue 63:
   https://abcde.einval.com/bugzilla/show_bug.cgi?id=63
 * Experimental support for embedding albumart with ogg files. The slightly
   tortuous technique drawn from 2 sources:

   1. https://github.com/biapy/howto.biapy.com/blob/master/various/mussync-tools
   2. https://github.com/acabal/scripts/blob/master/ogg-cover-art
   
   Testing is strongly encouraged, perhaps the simplest way to test is with:
   
       abcde -o ogg -B
    
    Or the appropriate settings in an ~/.abcde.conf file.
 * Massive rework of CD lookup code so support multiple sources
   better. Thanks to Gabriel Rosenkoetter for his initial idea in this
   area, and to Tomasz Goliński on irc for initial inspiration on how
   this should work better.
   There are now 3 different options for CD lookup: cddb, musicbrainz and
   cdtext. They can all be listed in a comma-separated list for
   CDDBMETHOD and the code will now call all of them in the sequence
   listed. All the results will be combined into one list at the end for
   the user to select, just like would have previously worked for one
   source only. Closes Issue 42:
   https://abcde.einval.com/bugzilla/show_bug.cgi?id=42
 * Fix Musicbrainz ID calculation in makeids()
   Apply fix suggested by petecollins24@gmail.com; add PREGAP to
   LEADOUT to correct Musicbrainz ID calculation. Hopefully closes
   Issue 54: https://abcde.einval.com/bugzilla/show_bug.cgi?id=54
 * Fix abcde.mkcue() when handling the --wholedisk option. Thanks to
   Nino Burini for the patch. Closes Issue 47 and maybe also 45:
   https://abcde.einval.com/bugzilla/show_bug.cgi?id=47
   https://abcde.einval.com/bugzilla/show_bug.cgi?id=45
 * Add more comprehensive examples for filename munging in the example
   config file. Closes Issue 49:
   https://abcde.einval.com/bugzilla/show_bug.cgi?id=49
 * Add a more better fix for the year lookup problem in abcde-musicbrainz-tool
   Thanks to Tom Samstag for the patch. Closes Issue 57:
   https://abcde.einval.com/bugzilla/show_bug.cgi?id=57
 * Redirect stderr on "which" calls to clear up error noise on some
   systems. Closes Issue 56:
   https://abcde.einval.com/bugzilla/show_bug.cgi?id=56
 * Add Irix support, based on a patch by abcde@canavan.de. Closes Issue 29:
   https://abcde.einval.com/bugzilla/show_bug.cgi?id=29
 * Rework abcde-musicbrainz-tool to work with WebService::MusicBrainz 1.x
   Thanks to Nicolas Guillaumin for the patch. Closes Issue 60:
   https://abcde.einval.com/bugzilla/show_bug.cgi?id=60
   Also added a specific dependency on version 1.0.4 or newer, so

abcde.py
#!/usr/bin/env python
# Copyright (c) 2007 Jesus Climent <jesus.climent@hispalinux.es>
# This code is hereby licensed for public consumption under either the
# GNU GPL v2 or greater, or Larry Wall's Artistic license - your choice.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

# import those needed modules.

import os
import re
import sys
import time
import getopt
import string
import select

__version__ = "1.0"

"""
abcde.py - A Better CD Encoder - python release
Copyright (C) 2007 Jesus Climent <jesus.climent@hispalinux.es>

"""

help = """This is abcde version """ + __version__ + """

usage: abcde.py [options] [tracks]
Options:
-1     Encode the whole CD in a single file
-a <action1[,action2]...>
       Actions to perform:
       cddb,read,normalize,encode,tag,move,replaygain,playlist,clean
-b     Enable batch normalization
-c <file>
       Specify a configuration file (overrides system and user config files)
-C <discid#>
       Specify discid to resume from (only needed if you no longer have the cd)
-d <device>
       Specify CDROM device to grab (flac uses a single-track flac file)
-D     Debugging mode (equivalent to sh -x abcde)
-e     Erase encoded track information from status file
-f     Force operations that otherwise are considered harmful. Read "man abcde"
-g     Use "lame --nogap" for MP3 encoding. Disables low disk and pipes flags
-h     This help information
#-i    Tag files while encoding, when possible (local only) -NWY-
-j <#> Number of encoder processes to run at once (localhost)
-k     Keep the wav tracks for later use
-l     Use low disk space algorithm
-L     Use local CDDB storage directory
-n     No lookup. Don't query CDDB, just create and use template
-N     Noninteractive. Never prompt for anything
-m     Modify playlist to include CRLF endings, to comply with some players
       WARNING: Deprecated. Use \"cue\" action
-o <type1[,type2]...>
       Output file type(s) (vorbis,mp3,flac,spx,mpc,wav,m4a). Defaults to vorbis
-p     Pad track numbers with 0's (if less than 10 tracks)
-P     Use UNIX pipes to read+encode without wav files
-q <level>
       Set quality level (high,medium,low)
-r <host1[,host2]...>
       Also encode on these remote hosts
-R     Use local CDDB in recursive mode
-s <field>
       Show dielfs from the CDDB info (year,genre)
-S <#> Set the CD speed
-t <#> Start the track numbering at a given number
-T <#> Same as -t but modifies tag numbering
-U     Do NOT use UNICODE (UTF8) tags and comments
-v     Show version number and exit
-V     Be a bit more verbose about what is happening behind the scenes
-x     Eject CD after all tracks are read
-w <comment>
       Add a comment to the CD tracks
-W <#> Concatenate CDs: -T #01 -w "CD #"
-z     Use debug CDROMREADERSYNTAX option (needs cdparanoia)

Tracks is a space-delimited list of tracks to grab.
Ranges specified with hyphens are allowed (i.e., 1-5).

"""

def usage ():
	print help

def addstatus(status,file):
	try:
		file = open(file, "w")
	except:
		log("error","file",file,"cannot be read")
		return -1

def log(status,logstring):
	if re.compile("info").match(status):
		status = "[INFO]"
		pass
	elif re.compile("warning").match(status):
		status = "[WARNING]"
		pass
	elif re.compile("error").match(status):
		returncode = 1
		status = "[ERROR] %s\n"
		sys.stderr.write(status % logstring)
		sys.exit(1)
	else:
		return 1

def f_seq_row (min,max):
	try:
		seq = range(min,max)
		return seq
	except:
		log(error,"syntax error while processing track numbers")
		return -1

def f_seq_line (min,max):
	try:
		seq = range(min,max)
		return seq
	except:
		log(error,"syntax error while processing track numbers")
		return -1

#usage()

# get_first and get_last can be substituted by range[0] and range[:-1]

# checkstatus(string)
# Returns "0" if the string was found, returns 1 if it wasn't
# Puts the blurb content, if available, on stdout.
# Otherwise, returns "".
def checkstatus (string, file):

	patern = re.compile("^"+string+"(=.*)?$")

	try:
		file = open(file, "r")
	except:
		log("error","file "+file+" cannot be read")
		return -1

	blurb = []
	while 1:
		line = file.readline()
		if line == "": break
		blurb.append(re.search(patern,line).string)

	print blurb
	if blurb[-1]:
		return 0
	else:
		return 1

# which(program)
# checks where we can find a program in the path
def which(program):
	for path in string.split(os.environ["PATH"], ":"):
		if os.path.exists(os.path.join(path, program)):
			return os.path.join(path, program)



def main():
	try:
		opts, args = getopt.getopt(sys.argv[1:], "1a:bc:C:d:DefghjklLnNmopPqrRsStTUvVxwWz")
	except:
		log("error","unknown error")
		sys.stderr.write(usage % sys.argv[0])
		sys.exit(1)

	try:
		#abcde.setup()
		for opt, optarg in opts:
			print opt, optarg
			if opt == "-1": o_onetrack = "y"
			if opt == "-a": o_actions = optarg
			if opt == "-b": o_batchnormalize = "y"
			if opt == "-B": o_nobatchreplygain = "y"
			if opt == "-c":
				o_configfile = str(optarg)
				try:
					if not re.compile("\.\/").search(o_configfile):
						o_configfile = os.environ.get("PWD", "./") + "/" + o_configfile
					os.path.exists(o_configfile)
				except:
					log("error",o_configfile+" cannot be read")
			if opt == "-C":
				if re.compile("abcde\.").match(optarg):
					o_discid = optarg
				else:
					o_discid = "abcde." + optarg
			if opt == "-d": o_cdrom = optarg
			if opt == "-D": o_debug = "y"
			if opt == "-h":
				usage()
				sys.exit(1)
			if opt == "-e": o_eraseencodedstatus = "y"
#			if opt == "-E": o_encoding
			if opt == "-f": o_force = "y"
			if opt == "-g": o_nogap = "y"
			if opt == "-j": o_maxprocs = optarg
			if opt == "-k": o_keepwavs = "y"
			if opt == "-l": o_lowdisk = "y"
			if opt == "-L": o_localcddb = "y"
			if opt == "-n": o_cddbavailable = "n"
			if opt == "-m": o_dosplaylist = "y"
			if opt == "-M": o_docue = "y"
			if opt == "-N": o_interactive = "n"
			if opt == "-o": o_outputtypes = optarg
			if opt == "-p": o_padtracks = "y"
			if opt == "-P": o_usepipes = "y"
			if opt == "-q": o_qualitylevel = optarg
			#if opt == "-r": o_remotehosts = optarg
			if opt == "-R": o_localcddbrecursive = "y"
			if opt == "-s": o_showcddbfields = "y"
			if opt == "-S": o_cdspeed = optarg
			if opt == "-t": o_starttracknumber = optarg
			if opt == "-T":
				o_starttracknumber = optarg
				o_starttracknumbertag = "y"
			if opt == "-U": o_cddbproto = 5
			if opt == "-v":
				print "This is abcde v", __version__
				print "Usage: abcde.py [options] [tracks]"
				print "abcde -h for extra help"
				sys.exit(0)
			if opt == "-V": o_verbose = "y"
			if opt == "-x": o_eject = "y"
#			if opt == "-X": o_cue2discid = optarg
			if opt == "-w": o_comment = optarg
			if opt == "-W":
				if re.compile("^\d+$").search(optarg)
					o_starttracknumber = optarg + "01"
					o_starttracknumbertag = "y"
					o_comment = "CD" + optarg
				else:
					log("error","opt -W must be an integer")
					sys.exit(1)

	except:
		#log("error","arguments were not correct")
		pass


#try:
##	checkstatus("test", "/tmp/status")
#	pass
#except:
#	sys.exit(1)

# -------------------------------
if __name__ == "__main__": main()

# b:is_python
# vim:tabstop=4:
